### ISSUE TITLE
Non-integer playback speed of sounds causes crackling

### ISSUE THREAD

**Issue by underwires** [defold/defold#5613]
**Describe the bug (REQUIRED)**
Whenever a sound is played back with a non-integer 

**To Reproduce**
Load and play a sound with a non-integer speed, e.g. 1.259
`sound.play("sounds#pluck-gritty-E3", {delay = 0, gain = 0.1, pan = 0.0, speed = 1.259})`

See attached WAV files.

**Expected behavior**
It should be pitched without audible artifacts

**Defold version:**
  - Version [2.0]

**Platforms:**
 - Platforms: [macOS]
 - OS: [Catalina 10.15.7]
 - Device: [MacBook Pro]

**WAV file examples**
[WAV-files.zip](https://github.com/defold/defold/files/6080192/WAV-files.zip)

Pluck Gritty E3 Original.wav - The original sound
Pluck Gritty E3 x1.259 Audacity.wav - the sound played at 1.259x speed in Audacity with no glitches
Pluck Gritty E3 x1.259 Defold.wav - the sound played at 1.259x speed in Defold with glitches

Glitches can be found at these time locations (Defold is on top):

0.0348 seconds
![defold sound speed glitch 1](https://user-images.githubusercontent.com/17102069/109895865-d9491d80-7c44-11eb-8216-fb317ec32f6c.png)

0.0523 seconds
![defold sound speed glitch 2](https://user-images.githubusercontent.com/17102069/109895874-dfd79500-7c44-11eb-8c6d-b71e58f96dfe.png)

0.0697 seconds
![defold sound speed glitch 3](https://user-images.githubusercontent.com/17102069/109895888-e239ef00-7c44-11eb-8b75-aecf3c825558.png)

0.0175 seconds
![defold sound speed glitch 4](https://user-images.githubusercontent.com/17102069/109895898-e49c4900-7c44-11eb-8907-874e5cc831a7.png)

**Comment by subsoap** [comment]
@underwires For reference do the crackling/glitches happen in release builds too?

**Comment by canted** [comment]
@subsoap it looks like the actual version I am using is 

### Version: 1.2.179 
### Channel: editor-alpha
### Build time: 2021-02-19T13:21:56.183041
### Editor SHA dc0725b3f70d22c8e0b1ad09bfd5f4f233d7fb0f
### Engine SHA 5209b50cf4bf8e925cb4c32fd0bde6c53fc85a66

I don't know where I got the 2.0, I thought it was an 'about defold' pop up window but now I can't find it.  Sorry about the confusion.

**Comment by subsoap** [comment]
@canted I mean if you bundle the game as a release build does the crackling still happen there.

**Comment by canted** [comment]
@subsoap yes, it happens in the bundle too

**Comment by otsakir** [comment]
So, it seems that the sound artifact appears every 0.0174 secs. That's the time increase between the four screenshots taken above.

Interestingly, the size of the buffer mixed in a go, is 767 frames. I got that from playing a little with gdb on the same sound files and settings. The time needed to play that at 44100Hz is:
767*1/44100 = 0.0173922 sec.

As expected, the glitch occurs _between_ mixing iterations.

**Comment by otsakir** [comment]
I submitted a fix for the issue. 

The problem stems from the fact that mixing uses the concept of fractional frames in defold. Which, once i understood, seemed pretty neat but also complicated. 

### Summary

When a sound is mixed into the mix buffer buffer at speed > 1, not all source samples are copied to the buffer. Some sort of selection should take place. And since the array indices are integer while the speed is float, there is some sort of float interval between the consecutive indices to source samples. The actual sample taken from the source array is typically a `mix` from _two_ samples and there is some sort of bias regarding how much each one will count (@JCash , that proved to be the reason for having a pair of values in mono sounds). This bias depends on the fractional part of the index, stored in `SoundInstance.m_FrameFraction`. Since mixing a sound instance is done in many steps, each one mixing a limited count of frames, the fractional part that is left from each iteration plays an important role for the algorithm. 

This issue, to my understanding was that a wrong number of samples were fetched here - https://github.com/defold/defold/blob/dev/engine/sound/src/sound.cpp#L1136 . As the result was a float number  and the fractional part was truncated. This resulted in fetching fewer frames than the algorithm needed to operate correctly. I used ceil() to make sure that the next frame is loaded.

**Comment by otsakir** [comment]
I investigated further. 

I restored old code so that i could here the crackling. I logged mixed samples to file in various parts of sound.cpp. At Master(), there were 2048 frames mixed in a chunk. But the last sample of the chunk was always 0 and clearly didn't follow the curve of the audio. It was a sharp change. Obviously the reason for the crackling.

In Mix() however, the `mix_count` was calculated as 2047. So, the last frame in the mix_buffer was not touched at all. And was still 0 when pushed to master.

The following piece of code calculates the number of frames to mix. It takes into account speed and sound rate change. Something is wrong here. Or at least not in accordance with the size calculations in MixInstance().

https://github.com/otsakir/defold/blob/d18a5e34780892bef749143198bb42ca1d987fbe/engine/sound/src/sound.cpp#L1074-L1076

It seems the reason why my initial patch worked is because the use of `ceil()` resulted in pushing one more frame in the mix buffer.

**Comment by otsakir** [comment]
Ok. Mystery solved. It's all about small rounding errors when dealing with frame calculations. Three different issues have been observed.

1. Small crackling heard between sound chunks.
2. Small crackling heard with looping enabled.
3. Worse crackling heard ~every 6 seconds  

### Case 1 & 2

Example. Lets say that we have a mix buffer with 10 frames and we mix a sound instance at speed 1.259. We'll need 10*1.259 = 12.59 frames. The fractional part is droped here when converting to an integer.

`uint32_t n = ceilf(sound->m_FrameCount * dmMath::Max(1.0f, instance->m_Speed) - instance->m_FrameCount); `

https://github.com/otsakir/defold/blob/d18a5e34780892bef749143198bb42ca1d987fbe/engine/sound/src/sound.cpp#L1145

Later, in Mix() we use this truncated number of instance frames to get back to the number of frames in the mix buffer. But we won't get a round 10 since we already chopped off the fractional part of the instance frame count. We'll fall back to the previous integer, 9. The last frame will keep its default value, 0.

Both of this issues are fixed when adding ceilf() to the calculation of the needed instance frame to load. 
NOTE: Theoritically, ceilf() adds a small quantity to the resulting size number that's still wrong since it could result in fetching more frames. However, appears to handle this case as it consumes the instance frames needed and leaves the rest intact.

### Case 3

There was another more severe anomaly observed in the audio when i let the sound play for more time. It occured every 6 seconds. Although i didn't investigate it in depth, it seemed fixed when i took into account the speed in a couple of places where sound->m_FrameCount was compared with instance->m_FrameCount. Still in MixInstance() . These fixes have not been commited yet.

**Comment by JCash** [comment]
Good investigation!

> An easy test would be to loop for a longer time and carefully...

Although we want correctness, we also don't want to spend too much time playing sounds. Our build+test times are already quite long. So use the "playing" strategy wisely.

I think in the future, if we detect more issues like this, we'd like to have a custom "decoder", that produces a deterministic sequence of values (e.g. a saw tooth wave) that's easy to verify that the values are within a good threshold when playing/looping at different speeds.

**Comment by otsakir** [comment]
> I think in the future, if we detect more issues like this, we'd like to have a custom "decoder", that produces a deterministic sequence of values (e.g. a saw tooth wave) that's easy to verify that the values are within a good threshold when playing/looping at different speeds.

Good idea. Currently, the automatically generated sounds in the tests follow a sinusoidal function i think. Not ideal, but you can roughly follow the pattern and detect anomalies.
